src/main/java/minijava/token/Token.java

```diff
@@ -1,50 +1,59 @@
 package minijava.token;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static minijava.token.Terminal.Associativity;
+
+import java.util.Arrays;
+import org.jetbrains.annotations.Nullable;
+
 /** Instances of this class are immutable. */
 public class Token {
 
   public final Terminal terminal;
   public final Position position;
-  public final String lexval;
-
-  public Token(Terminal terminal, Position position, String lexval) {
-    this.terminal = terminal;
-    this.position = position;
-    this.lexval = lexval.intern();
-  }
+  @Nullable public final String lexval;
 
-  public boolean isTerminal(Terminal otherTerminal) {
-    return terminal.equals(otherTerminal);
+  public Token(Terminal terminal, Position position, @Nullable String lexval) {
+    this.terminal = checkNotNull(terminal);
+    this.position = checkNotNull(position);
+    this.lexval = lexval == null ? null : lexval.intern();
   }
 
-  public boolean isEOF() {
-    return terminal.equals(Terminal.EOF);
+  public boolean isOperator() {
+    return terminal.associativity != null;
   }
 
-  public boolean isType(Terminal.TerminalType type) {
-    return terminal.isType(type);
-  }
-
-  @Override
-  public String toString() {
```

Comment:
I'm not sure what you mean by that. Do you need a method that outputs something different from the `toString` method?